pinot-minion-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/CustomerBasedRetentionTaskGenerator.java

```diff
@@ -0,0 +1,203 @@
+            .warn("Found incomplete tasks: {} for same table: {}. Skipping task generation.", incompleteTasks.keySet(),
+                offlineTableName);
+        continue;
+      }
+
+      TableTaskConfig tableTaskConfig = tableConfig.getTaskConfig();
+      Preconditions.checkState(tableTaskConfig != null);
+      Map<String, String> taskConfigs = tableTaskConfig.getConfigsForTaskType(TASK_TYPE);
+      Preconditions.checkState(taskConfigs != null, "Task config shouldn't be null for table: {}", offlineTableName);
+
+      // Get customer retention config
+      Map<String ,String> customerRetentionConfigMap = getCustomerRetentionConfig();
+      String customerRetentionConfigMapString = customerRetentionConfigMap.keySet().stream()
+          .map(key -> key + "=" + customerRetentionConfigMap.get(key))
+          .collect(Collectors.joining(", ", "{", "}"));
+      Set<String> distinctRetentionPeriods = getDistinctRetentionPeriods(customerRetentionConfigMap);
+
+      /**
+       * Generate one task per retention period.
+       * This is because we have to update watermarks based on retention period.
+       */
+      for (String retentionPeriod : distinctRetentionPeriods) {
+
+        // Get the bucket size
+        String bucketTimePeriod = taskConfigs.getOrDefault(BUCKET_TIME_PERIOD_KEY, retentionPeriod);
+        long bucketMs = TimeUtils.convertPeriodToMillis(bucketTimePeriod);
+
+        // Get watermark from OfflineSegmentsMetadata ZNode. WindowStart = watermark. WindowEnd = windowStart + bucket.
+        long windowStartMs = getWatermarkMs(offlineTableName, bucketMs);
+        long windowEndMs = windowStartMs + bucketMs;
+
+        List<String> segmentNames = new ArrayList<>();
+        List<String> downloadURLs = new ArrayList<>();
+
+        for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : _clusterInfoAccessor.getOfflineSegmentsMetadata(offlineTableName)) {
+          // Only submit segments that have not been converted
+          Map<String, String> customMap = offlineSegmentZKMetadata.getCustomMap();
+          if (customMap == null || !customMap.containsKey(
+              MinionConstants.ConvertToRawIndexTask.COLUMNS_TO_CONVERT_KEY + MinionConstants.TASK_TIME_SUFFIX)) {
+            segmentNames.add(offlineSegmentZKMetadata.getSegmentName());
+            downloadURLs.add(offlineSegmentZKMetadata.getDownloadUrl());
+          }
+        }
+
+        if (!segmentNames.isEmpty()) {
+          Map<String, String> configs = new HashMap<>();
+          configs.put(MinionConstants.TABLE_NAME_KEY, offlineTableName);
+          configs.put(MinionConstants.SEGMENT_NAME_KEY, StringUtils.join(segmentNames, ","));
+          configs.put(MinionConstants.DOWNLOAD_URL_KEY, StringUtils.join(downloadURLs, MinionConstants.URL_SEPARATOR));
+          configs.put(MinionConstants.UPLOAD_URL_KEY, _clusterInfoAccessor.getVipUrl() + "/segments");
+
+          configs.put(CUSTOMER_RETENTION_CONFIG, customerRetentionConfigMapString);
```

Comment:
@laxmanchekka do we need this ?

Reply:
Yes. To decide whether a record needs to be dropped, we need this config